### PR TITLE
Merge roi head modules

### PR DESCRIPTION
**Pull request by vsuthichai**

### CODE REVIEW

**Comment by vsuthichai** [comment]
The no batch else block still uses BoxProposals, and the batch block still uses it in once place but this is just to accommodate for the mask blocks which still use BoxProposals

File: MaskRCNN_no_batch_convergence/train.py [hunk 1]
```diff
@@ -363,26 +364,23 @@ def rpn(self, image, features, inputs):
...
         gt_boxes, gt_labels, *_ = targets
 
-        if self.training:
-
-
-
+        #############################################################################################################
+        if BATCH_FRCNN_HEAD:
+        #############################################################################################################
+            prepadding_gt_counts = tf.expand_dims(tf.shape(gt_labels)[0], axis=0)  # 1 x NumGT
```
Comment (vsuthichai) [inline MaskRCNN_no_batch_convergence/train.py:377]: I've moved this here because it was being used in two blocks: BATCH_SAMPLE_TARGETS and BATCH_FAST_RCNN_LOSSES

File: MaskRCNN_no_batch_convergence/train.py [hunk 2]
```diff
@@ -391,56 +389,29 @@ def roi_heads(self, image, features, proposals, targets):
...
-
-
-
-
+                proposal_boxes = proposal_boxes[:, 1:]
 
+                # TODO: Remove this after mask blocks have been modified to work with no BoxProposals
+                # It's needed for mask block functionality but is not used within this code block and
+                # should be able to be safely deleted once the mask blocks have been converted.
+                proposals = BoxProposals(proposal_boxes, proposal_labels, proposal_gt_id_for_each_fg[0])
```
Comment (vsuthichai) [inline MaskRCNN_no_batch_convergence/train.py:397]: Delete this once the mask blocks have been refactored to not use BoxProposals.

File: MaskRCNN_no_batch_convergence/train.py [hunk 3]
```diff
@@ -452,35 +423,46 @@ def roi_heads(self, image, features, proposals, targets):
...
+                all_losses = fastrcnn_head.losses(BATCH_SIZE_PLACEHOLDER)
 
-            fastrcnn_head = FastRCNNHead(proposals, fastrcnn_box_logits, fastrcnn_label_logits,
-                                     gt_boxes, tf.constant(cfg.FRCNN.BBOX_REG_WEIGHTS, dtype=tf.float32))
-            if self.training:
-                all_losses = fastrcnn_head.losses()
-        ##########################################################################################################
+        else: # BATCH_FRCNN_HEAD = False
 
+            proposals = BoxProposals(proposal_boxes)
```
Comment (vsuthichai) [inline MaskRCNN_no_batch_convergence/train.py:444]: `self.rpn` will return proposal_boxes now instead of `BoxProposals`.  Need to move the constructor further down so the no batch code path still works.

**Comment by armandmcqueen** [comment]
I prefer calling it BoxClass head as opposed to frcnn head, but right now frcnn makes sense given the way files are named.

Looks good. Approved pending convergence test results.

**Comment by vsuthichai** [comment]
Grey is with flag set to True, the rest are past converging runs. 

Screen Shot 2019-03-28 at 9 41 58 AM

Screen Shot 2019-03-28 at 9 42 46 AM

**Comment by vsuthichai** [comment]
Grey is with flag set to False, the rest are past converging runs

Screen Shot 2019-03-28 at 9 44 52 AM

Screen Shot 2019-03-28 at 9 45 41 AM
